nvme.c

```diff
@@ -5317,6 +5398,629 @@ static int show_registers(int argc, char **argv, struct command *cmd, struct plu
+			flag = true;
+		break;
+	case NVME_REG_PMRCTL:
+		if (cfg->pmrctl)
+			flag = true;
+		break;
+	case NVME_REG_PMRSTS:
+		if (cfg->pmrsts)
+			flag = true;
+		break;
+	case NVME_REG_PMREBS:
+		if (cfg->pmrebs)
+			flag = true;
+		break;
+	case NVME_REG_PMRSWTP:
+		if (cfg->pmrswtp)
+			flag = true;
+		break;
+	case NVME_REG_PMRMSCL:
+		if (cfg->pmrmscl)
+			flag = true;
+		break;
+	case NVME_REG_PMRMSCU:
+		if (cfg->pmrmscu)
+			flag = true;
+		break;
+	default:
+		break;
+	}
+
+	if (flag)
+		nvme_show_ctrl_register(bar, fabrics, offset, flags);
+
+	return flag;
+}
+
+static bool get_register_offset(void *bar, bool fabrics, struct get_reg_config *cfg,
+				enum nvme_print_flags flags)
+{
+	bool flag = cfg->offset >= 0;
+
+	if (flag)
+		nvme_show_ctrl_register(bar, fabrics, cfg->offset, flags);
+
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_CAP, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_VS, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_INTMS, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_INTMC, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_CC, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_CSTS, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_NSSR, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_AQA, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_ASQ, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_ACQ, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_CMBLOC, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_CMBSZ, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_BPINFO, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_BPRSEL, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_BPMBL, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_CMBMSC, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_CMBSTS, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_CMBEBS, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_CMBSWTP, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_NSSD, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_CRTO, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_PMRCAP, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_PMRCTL, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_PMRSTS, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_PMREBS, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_PMRSWTP, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_PMRMSCL, flags) || flag;
+	flag = get_register_flag(bar, fabrics, cfg, NVME_REG_PMRMSCU, flags) || flag;
```

Comment:
This logic is hard to follow. I think it would be clearer to have a single helper function to compute the offset from the flags:
```c
int get_register_offset(const struct get_reg_config *cfg)
{
	int offset = cfg->offset;
	if (cfg->cap) {
		if (offset >= 0)
			return -1; // multiple offsets specified
		offset = NVME_REG_CAP;
	}
	if (cfg->vs) {
		if (offset >= 0)
			return -1; // multiple offsets specified
		offset = NVME_REG_VS;
	}
	// ...
	return offset;
}
```
I also don't think this is enforcing exclusivity. If multiple flags are set, they will all get printed. This may work for PCIe, where the whole bar is mapped into memory, but definitely doesn't for fabrics.

Reply:
Just refactored the function to be more simple. For the fabrics works for the some registers CAP, VS, CC, CSTS, NSSR and CRTO to print multiply.